# Limpieza de la tabla de DIM_TIENDA_TEST

---

### Importación de librerías

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.impute import KNNImputer
from sklearn.preprocessing import LabelEncoder, StandardScaler

### Importación de datos

In [2]:
ruta = "C:/Users/danyj/OneDrive/Documents/Datathon/Raw Data"

In [3]:
dim_tienda_test = pd.read_csv(f"{ruta}/DIM_TIENDA_TEST.csv")
dim_tienda_test

TIENDA_ID  PLAZA_CVE NIVELSOCIOECONOMICO_DES ENTORNO_DES  MTS2VENTAS_NUM  \
0          680          1                       C       Hogar          102.36   
1          730          1                       C       Hogar           97.43   
2          650          1                       C       Hogar          117.01   
3          670          1                       C        Base          109.76   
4          800          1                       C    Peatonal            0.00   
..         ...        ...                     ...         ...             ...   
100        720          6                       C       Hogar            0.00   
101        130          6                      CD        Base          110.06   
102        320          6                       C       Hogar          122.00   
103        510          6                       B       Hogar            0.00   
104        500          6                       B       Hogar            0.00   

     PUERTASREFRIG_NUM  CAJONESESTACIONAMIENTO_NUM  LATITUD_NUM  LONGITUD_NUM  \
0                   13                           0     25.65488    -100.21207   
1                   14                           0     25.66315    -100.22738   
2                   13                           0     25.66404    -100.22993   
3                   13                           0     25.66508    -100.26338   
4                    0                           0     25.69367    -100.21433   
..                 ...                         ...          ...           ...   
100                  0                           0     25.60253     -99.99317   
101                 10                           0     25.36439    -100.10202   
102                 10                           0     25.57214    -100.01371   
103                 13                           0     25.19747     -99.83198   
104                 13                           0     25.60256    -100.08093   

    SEGMENTO_MAESTRO_DESC  LID_UBICACION_TIENDA DATASET  
0           Hogar Reunión           UT_DENSIDAD    TEST  
1           Hogar Reunión           UT_DENSIDAD    TEST  
2           Hogar Reunión           UT_DENSIDAD    TEST  
3           Hogar Reunión           UT_DENSIDAD    TEST  
4          Parada Técnica   UT_TRAFICO_PEATONAL    TEST  
..                    ...                   ...     ...  
100         Hogar Reunión           UT_DENSIDAD    TEST  
101         Hogar Reunión  UT_TRAFICO_VEHICULAR    TEST  
102         Hogar Reunión  UT_TRAFICO_VEHICULAR    TEST  
103         Hogar Reunión  UT_TRAFICO_VEHICULAR    TEST  
104         Hogar Reunión           UT_DENSIDAD    TEST  

[105 rows x 12 columns]

In [4]:
venta = pd.read_csv(f"{ruta}/Venta.csv")
venta

TIENDA_ID  MES_ID  VENTA_TOTAL
0            813  202404   1042761.16
1            742  202404    604946.96
2            436  202404   2486787.81
3            732  202404    680701.78
4            282  202404   1227768.82
...          ...     ...          ...
21376        885  202312   1157010.83
21377        793  202401    699916.91
21378         53  202401   1008108.28
21379        960  202401   1324156.07
21380        313  202401    626345.23

[21381 rows x 3 columns]

In [5]:
meta_venta = pd.read_csv(f"{ruta}/Meta_venta.csv")
meta_venta

ENTORNO_DES  Meta_venta
0        Base      480000
1       Hogar      490000
2    Peatonal      420000
3      Receso      516000

## Separación por los 105 IDs en la tabla de 'Venta'
---

Analizando los datos, nos dimos cuenta que los ID que estaban en 'Venta.csv' pero no en 'DIM_TIENDA.csv' son aquellos que se separaron para usarlos como TEST. Es por eso que se utilizarán esos para recrear el mismo proceso que se hizo para la limpieza de los datos de entrenamiento.

In [7]:
ids_test = [10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110,
            120, 130, 140, 150, 160, 170, 180, 190, 200, 210,
            220, 230, 240, 250, 260, 270, 280, 290, 300, 310,
            320, 330, 340, 350, 360, 370, 380, 390, 400, 410,
            420, 430, 440, 450, 460, 470, 480, 490, 500, 510,
            520, 530, 540, 550, 560, 570, 580, 590, 600, 610,
            620, 630, 640, 650, 660, 670, 680, 690, 700, 710,
            720, 730, 740, 750, 760, 770, 780, 790, 800, 810,
            820, 830, 840, 850, 860, 870, 880, 890, 900, 910,
            920, 930, 940, 950, 960, 970, 980, 990, 1000, 1010,
            1020, 1030, 1040, 1050]

venta = venta[venta['TIENDA_ID'].isin(ids_test)]

In [9]:
venta['TIENDA_ID'].nunique()

105

## Merge de las tablas

---

In [10]:
df_test = venta.merge(dim_tienda_test, on = 'TIENDA_ID', how = 'left')
df_test = df_test.merge(meta_venta, on = 'ENTORNO_DES', how = 'left')

# Creación de la columna 'éxito'

df_test["EXITO"] = (df_test["VENTA_TOTAL"] >= df_test["Meta_venta"]).astype(int)

In [11]:
df_test

TIENDA_ID  MES_ID  VENTA_TOTAL  PLAZA_CVE NIVELSOCIOECONOMICO_DES  \
0           250  202404   2019799.45          5                       C   
1           970  202404    959735.17          2                       C   
2           400  202404    831299.01          5                       C   
3           450  202404   1208747.34          5                       C   
4           380  202405   1178024.41          1                       C   
...         ...     ...          ...        ...                     ...   
2123       1000  202401   1029569.92          2                       C   
2124        640  202401    668567.17          5                       C   
2125        850  202401      6505.79          4                       C   
2126        560  202312   1624049.08          1                       C   
2127        960  202401   1324156.07          2                      BC   

     ENTORNO_DES  MTS2VENTAS_NUM  PUERTASREFRIG_NUM  \
0          Hogar          118.93                 13   
1       Peatonal           83.47                  7   
2          Hogar          112.20                 13   
3          Hogar          105.24                 13   
4           Base          101.66                 13   
...          ...             ...                ...   
2123       Hogar          101.82                 12   
2124       Hogar          109.90                 13   
2125    Peatonal          118.65                 14   
2126       Hogar          112.33                 13   
2127       Hogar          122.00                 10   

      CAJONESESTACIONAMIENTO_NUM  LATITUD_NUM  LONGITUD_NUM  \
0                              6     25.76580    -100.23715   
1                              0     26.09231     -98.27756   
2                              7     25.82921    -100.11680   
3                              9     25.81408    -100.08988   
4                             10     25.70768    -100.17905   
...                          ...          ...           ...   
2123                          10     26.01994     -98.28429   
2124                           0     25.78546    -100.18814   
2125                           0     25.69801    -100.10738   
2126                           6     25.69058    -100.22989   
2127                          10     26.07600     -98.35906   

     SEGMENTO_MAESTRO_DESC  LID_UBICACION_TIENDA DATASET  Meta_venta  EXITO  
0            Hogar Reunión           UT_DENSIDAD    TEST      490000      1  
1              Oficinistas   UT_TRAFICO_PEATONAL    TEST      420000      1  
2            Hogar Reunión           UT_DENSIDAD    TEST      490000      1  
3           Parada Técnica           UT_DENSIDAD    TEST      490000      1  
4              Oficinistas         UT_GAS_URBANA    TEST      480000      1  
...                    ...                   ...     ...         ...    ...  
2123         Hogar Reunión           UT_DENSIDAD    TEST      490000      1  
2124         Hogar Reunión           UT_DENSIDAD    TEST      490000      1  
2125        Parada Técnica  UT_TRAFICO_VEHICULAR    TEST      420000      0  
2126         Hogar Reunión  UT_TRAFICO_VEHICULAR    TEST      490000      1  
2127         Hogar Reunión           UT_DENSIDAD    TEST      490000      1  

[2128 rows x 16 columns]

## Análisis y procesamiento 

---

In [12]:
# Dimensión del dataset
print(f"Número de filas: {df_test.shape[0]}")
print(f"Número de columnas: {df_test.shape[1]}")

Número de filas: 2128
Número de columnas: 16


In [14]:
# Valores faltantes por columna
df_test.isna().sum()

TIENDA_ID                     0
MES_ID                        0
VENTA_TOTAL                   0
PLAZA_CVE                     0
NIVELSOCIOECONOMICO_DES       0
ENTORNO_DES                   0
MTS2VENTAS_NUM                0
PUERTASREFRIG_NUM             0
CAJONESESTACIONAMIENTO_NUM    0
LATITUD_NUM                   0
LONGITUD_NUM                  0
SEGMENTO_MAESTRO_DESC         0
LID_UBICACION_TIENDA          0
DATASET                       0
Meta_venta                    0
EXITO                         0
dtype: int64

In [15]:
# Ver cantidad de 0s que hay en columnas que no deberían tener 0s

df_test[df_test['VENTA_TOTAL'] == 0]

Empty DataFrame
Columns: [TIENDA_ID, MES_ID, VENTA_TOTAL, PLAZA_CVE, NIVELSOCIOECONOMICO_DES, ENTORNO_DES, MTS2VENTAS_NUM, PUERTASREFRIG_NUM, CAJONESESTACIONAMIENTO_NUM, LATITUD_NUM, LONGITUD_NUM, SEGMENTO_MAESTRO_DESC, LID_UBICACION_TIENDA, DATASET, Meta_venta, EXITO]
Index: []

In [17]:
df_test[df_test['MTS2VENTAS_NUM'] == 0]

TIENDA_ID  MES_ID  VENTA_TOTAL  PLAZA_CVE NIVELSOCIOECONOMICO_DES  \
9            10  202405    669062.13          2                       C   
13          740  202412    356391.27          6                       C   
19          710  202405    973969.62          4                       C   
20          630  202412    348280.55          3                       C   
30          460  202407   1635951.92          3                       C   
...         ...     ...          ...        ...                     ...   
2102        590  202309   1045091.87          4                       C   
2108        710  202310    593275.56          4                       C   
2115        660  202311    477480.73          4                       B   
2120        460  202401   1533770.23          3                       C   
2121        780  202401   1028726.82          6                       C   

     ENTORNO_DES  MTS2VENTAS_NUM  PUERTASREFRIG_NUM  \
9           Base             0.0                  0   
13         Hogar             0.0                  0   
19         Hogar             0.0                  0   
20          Base             0.0                  0   
30         Hogar             0.0                 13   
...          ...             ...                ...   
2102       Hogar             0.0                 13   
2108       Hogar             0.0                  0   
2115        Base             0.0                  0   
2120       Hogar             0.0                 13   
2121      Receso             0.0                  0   

      CAJONESESTACIONAMIENTO_NUM  LATITUD_NUM  LONGITUD_NUM  \
9                              0     26.02203     -98.27759   
13                             0     25.59820     -99.99200   
19                             0     25.61985    -100.08667   
20                             0     25.66690    -100.36854   
30                             0     25.61702    -100.26720   
...                          ...          ...           ...   
2102                           0     25.66311    -100.09819   
2108                           0     25.61985    -100.08667   
2115                           0     25.64418    -100.08754   
2120                           0     25.61702    -100.26720   
2121                           0     25.51353    -100.19835   

     SEGMENTO_MAESTRO_DESC  LID_UBICACION_TIENDA DATASET  Meta_venta  EXITO  
9            Hogar Reunión  UT_TRAFICO_VEHICULAR    TEST      480000      1  
13           Hogar Reunión           UT_DENSIDAD    TEST      490000      0  
19           Hogar Reunión           UT_DENSIDAD    TEST      490000      1  
20             Oficinistas  UT_TRAFICO_VEHICULAR    TEST      480000      0  
30           Hogar Reunión           UT_DENSIDAD    TEST      490000      1  
...                    ...                   ...     ...         ...    ...  
2102         Hogar Reunión           UT_DENSIDAD    TEST      490000      1  
2108         Hogar Reunión           UT_DENSIDAD    TEST      490000      1  
2115           Oficinistas  UT_TRAFICO_VEHICULAR    TEST      480000      0  
2120         Hogar Reunión           UT_DENSIDAD    TEST      490000      1  
2121        Parada Técnica  UT_TRAFICO_VEHICULAR    TEST      516000      1  

[579 rows x 16 columns]

In [19]:
df_test[df_test['MTS2VENTAS_NUM'] == 0]['TIENDA_ID'].nunique()

30

### Imputación

In [20]:
# Crear dataset único por tienda
tiendas_unicas = df_test.drop_duplicates(subset=['TIENDA_ID']).copy()

# Agregar promedio de ventas por tienda
ventas_promedio = df_test.groupby('TIENDA_ID')['VENTA_TOTAL'].mean().reset_index()
ventas_promedio.columns = ['TIENDA_ID', 'VENTA_AMT_promedio']
tiendas_unicas = tiendas_unicas.merge(ventas_promedio, on = 'TIENDA_ID', how = 'left')

df_knn = tiendas_unicas.copy()

# Encoding de categóricas
categoricas = ['PLAZA_CVE', 'NIVELSOCIOECONOMICO_DES', 'ENTORNO_DES', 'SEGMENTO_MAESTRO_DESC', 'LID_UBICACION_TIENDA']
le_dict = {}

for col in categoricas:
   le = LabelEncoder()
   df_knn[col + '_encoded'] = le.fit_transform(df_knn[col].astype(str))
   le_dict[col] = le

# Features para KNN
features_knn = ['PUERTASREFRIG_NUM', 'CAJONESESTACIONAMIENTO_NUM', 'VENTA_AMT_promedio',
               'PLAZA_CVE_encoded', 'NIVELSOCIOECONOMICO_DES_encoded', 'ENTORNO_DES_encoded', 
               'SEGMENTO_MAESTRO_DESC_encoded', 'LID_UBICACION_TIENDA_encoded']

# Matriz para KNN (incluye la columna target)
matriz_knn = df_knn[features_knn + ['MTS2VENTAS_NUM']].copy()

# Reemplazar 0s por NaN en la columna target
matriz_knn.loc[matriz_knn['MTS2VENTAS_NUM'] == 0, 'MTS2VENTAS_NUM'] = np.nan

# Escalar features numéricas
scaler = StandardScaler()
cols_numericas = ['PUERTASREFRIG_NUM', 'CAJONESESTACIONAMIENTO_NUM', 'VENTA_AMT_promedio']
matriz_knn[cols_numericas] = scaler.fit_transform(matriz_knn[cols_numericas])

# Aplicar KNN
knn_imputer = KNNImputer(n_neighbors=5, weights='distance')
matriz_imputada = knn_imputer.fit_transform(matriz_knn)

# Obtener valores imputados
df_knn['MTS2VENTAS_NUM_knn'] = matriz_imputada[:, -1]

# Identificar tiendas que necesitan imputación
tiendas_cero = df_knn[df_knn['MTS2VENTAS_NUM'] == 0].copy()

# Validación híbrida por grupos
for idx in tiendas_cero.index:
   tienda_actual = df_knn.loc[idx]
   valor_knn = tienda_actual['MTS2VENTAS_NUM_knn']
   
   # Buscar grupo similar (mismo segmento y ubicación)
   grupo = df_knn[
       (df_knn['SEGMENTO_MAESTRO_DESC'] == tienda_actual['SEGMENTO_MAESTRO_DESC']) &
       (df_knn['LID_UBICACION_TIENDA'] == tienda_actual['LID_UBICACION_TIENDA']) &
       (df_knn['MTS2VENTAS_NUM'] > 0)
   ]
   
   if len(grupo) >= 3:
       q25 = grupo['MTS2VENTAS_NUM'].quantile(0.25)
       q75 = grupo['MTS2VENTAS_NUM'].quantile(0.75)
       
       # Si KNN está fuera del rango, usar mediana del grupo
       if not (q25 <= valor_knn <= q75):
           valor_final = grupo['MTS2VENTAS_NUM'].median()
       else:
           valor_final = valor_knn
   else:
       # Si no hay suficientes datos del grupo, usar KNN
       valor_final = valor_knn
   
   df_knn.loc[idx, 'MTS2VENTAS_NUM'] = valor_final

# Diccionario de mapeo
mapeo_areas = df_knn.set_index('TIENDA_ID')['MTS2VENTAS_NUM'].to_dict()

# Aplicar a dataset completo
df_test['MTS2VENTAS_NUM'] = df_test['TIENDA_ID'].map(mapeo_areas)


In [22]:
df_test[df_test['MTS2VENTAS_NUM'] == 0]

Empty DataFrame
Columns: [TIENDA_ID, MES_ID, VENTA_TOTAL, PLAZA_CVE, NIVELSOCIOECONOMICO_DES, ENTORNO_DES, MTS2VENTAS_NUM, PUERTASREFRIG_NUM, CAJONESESTACIONAMIENTO_NUM, LATITUD_NUM, LONGITUD_NUM, SEGMENTO_MAESTRO_DESC, LID_UBICACION_TIENDA, DATASET, Meta_venta, EXITO]
Index: []

In [24]:
df_test[df_test['PUERTASREFRIG_NUM'] == 0]

TIENDA_ID  MES_ID  VENTA_TOTAL  PLAZA_CVE NIVELSOCIOECONOMICO_DES  \
9            10  202405    669062.13          2                       C   
13          740  202412    356391.27          6                       C   
19          710  202405    973969.62          4                       C   
20          630  202412    348280.55          3                       C   
35          660  202403    550085.83          4                       B   
...         ...     ...          ...        ...                     ...   
2095       1040  202307   1261739.73          2                       C   
2107       1030  202310    610065.25          2                       C   
2108        710  202310    593275.56          4                       C   
2115        660  202311    477480.73          4                       B   
2121        780  202401   1028726.82          6                       C   

     ENTORNO_DES  MTS2VENTAS_NUM  PUERTASREFRIG_NUM  \
9           Base      108.006345                  0   
13         Hogar      112.566915                  0   
19         Hogar      108.695439                  0   
20          Base      109.297885                  0   
35          Base      103.515790                  0   
...          ...             ...                ...   
2095       Hogar      104.314610                  0   
2107      Receso      121.300000                  0   
2108       Hogar      108.695439                  0   
2115        Base      103.515790                  0   
2121      Receso      108.031541                  0   

      CAJONESESTACIONAMIENTO_NUM  LATITUD_NUM  LONGITUD_NUM  \
9                              0     26.02203     -98.27759   
13                             0     25.59820     -99.99200   
19                             0     25.61985    -100.08667   
20                             0     25.66690    -100.36854   
35                             0     25.64418    -100.08754   
...                          ...          ...           ...   
2095                           0     26.22708     -98.59373   
2107                          13     26.40152     -99.02246   
2108                           0     25.61985    -100.08667   
2115                           0     25.64418    -100.08754   
2121                           0     25.51353    -100.19835   

     SEGMENTO_MAESTRO_DESC  LID_UBICACION_TIENDA DATASET  Meta_venta  EXITO  
9            Hogar Reunión  UT_TRAFICO_VEHICULAR    TEST      480000      1  
13           Hogar Reunión           UT_DENSIDAD    TEST      490000      0  
19           Hogar Reunión           UT_DENSIDAD    TEST      490000      1  
20             Oficinistas  UT_TRAFICO_VEHICULAR    TEST      480000      0  
35             Oficinistas  UT_TRAFICO_VEHICULAR    TEST      480000      1  
...                    ...                   ...     ...         ...    ...  
2095         Hogar Reunión  UT_TRAFICO_VEHICULAR    TEST      490000      1  
2107      Barrio Competido   UT_TRAFICO_PEATONAL    TEST      516000      1  
2108         Hogar Reunión           UT_DENSIDAD    TEST      490000      1  
2115           Oficinistas  UT_TRAFICO_VEHICULAR    TEST      480000      0  
2121        Parada Técnica  UT_TRAFICO_VEHICULAR    TEST      516000      1  

[411 rows x 16 columns]

In [26]:
df_test[df_test['PUERTASREFRIG_NUM'] == 0]['TIENDA_ID'].nunique()

22

In [28]:
# Dataset único por tienda 
tiendas_unicas = df_test.drop_duplicates(subset=['TIENDA_ID']).copy()

# Agregar promedio de ventas por tienda
ventas_promedio = df_test.groupby('TIENDA_ID')['VENTA_TOTAL'].mean().reset_index()
ventas_promedio.columns = ['TIENDA_ID', 'VENTA_AMT_promedio']
tiendas_unicas = tiendas_unicas.merge(ventas_promedio, on = 'TIENDA_ID', how = 'left')

# Preparar datos para KNN
df_knn = tiendas_unicas.copy()

# Encoding de categóricas
categoricas = ['PLAZA_CVE', 'NIVELSOCIOECONOMICO_DES', 'ENTORNO_DES', 'SEGMENTO_MAESTRO_DESC', 'LID_UBICACION_TIENDA']
le_dict = {}

for col in categoricas:
    le = LabelEncoder()
    df_knn[col + '_encoded'] = le.fit_transform(df_knn[col].astype(str))
    le_dict[col] = le

# Features para KNN (ahora MTS2VENTAS_NUM es predictor, no target)
features_knn = ['MTS2VENTAS_NUM', 'CAJONESESTACIONAMIENTO_NUM', 'VENTA_AMT_promedio',
                'PLAZA_CVE_encoded', 'NIVELSOCIOECONOMICO_DES_encoded', 'ENTORNO_DES_encoded', 
                'SEGMENTO_MAESTRO_DESC_encoded', 'LID_UBICACION_TIENDA_encoded']

# Crear matriz para KNN
matriz_knn = df_knn[features_knn + ['PUERTASREFRIG_NUM']].copy()

# Reemplazar 0s por NaN en PUERTASREFRIG_NUM
matriz_knn.loc[matriz_knn['PUERTASREFRIG_NUM'] == 0, 'PUERTASREFRIG_NUM'] = np.nan

# Escalar features numéricas
scaler = StandardScaler()
cols_numericas = ['MTS2VENTAS_NUM', 'CAJONESESTACIONAMIENTO_NUM', 'VENTA_AMT_promedio']
matriz_knn[cols_numericas] = scaler.fit_transform(matriz_knn[cols_numericas])

# Aplicar KNN
knn_imputer = KNNImputer(n_neighbors=5, weights='distance')
matriz_imputada = knn_imputer.fit_transform(matriz_knn)

# Obtener valores imputados
df_knn['PUERTASREFRIG_NUM_knn'] = matriz_imputada[:, -1]

# Identificar tiendas que necesitan imputación
tiendas_cero_puertas = df_knn[df_knn['PUERTASREFRIG_NUM'] == 0].copy()

# Validación por grupos
for idx in tiendas_cero_puertas.index:
    tienda_actual = df_knn.loc[idx]
    valor_knn = tienda_actual['PUERTASREFRIG_NUM_knn']
    
    grupo = df_knn[
        (df_knn['SEGMENTO_MAESTRO_DESC'] == tienda_actual['SEGMENTO_MAESTRO_DESC']) &
        (df_knn['LID_UBICACION_TIENDA'] == tienda_actual['LID_UBICACION_TIENDA']) &
        (df_knn['PUERTASREFRIG_NUM'] > 0)
    ]
    
    if len(grupo) >= 3:
        q25 = grupo['PUERTASREFRIG_NUM'].quantile(0.25)
        q75 = grupo['PUERTASREFRIG_NUM'].quantile(0.75)
        
        if not (q25 <= valor_knn <= q75):
            valor_final = grupo['PUERTASREFRIG_NUM'].median()
        else:
            valor_final = valor_knn
    else:
        valor_final = valor_knn
    
    # Redondear a entero (las puertas son números enteros)
    df_knn.loc[idx, 'PUERTASREFRIG_NUM'] = round(valor_final)

# Crear diccionario de mapeo
mapeo_puertas = df_knn.set_index('TIENDA_ID')['PUERTASREFRIG_NUM'].to_dict()

# Aplicar a dataset completo
df_test['PUERTASREFRIG_NUM'] = df_test['TIENDA_ID'].map(mapeo_puertas)


In [30]:
df_test[df_test['PUERTASREFRIG_NUM'] == 0]

Empty DataFrame
Columns: [TIENDA_ID, MES_ID, VENTA_TOTAL, PLAZA_CVE, NIVELSOCIOECONOMICO_DES, ENTORNO_DES, MTS2VENTAS_NUM, PUERTASREFRIG_NUM, CAJONESESTACIONAMIENTO_NUM, LATITUD_NUM, LONGITUD_NUM, SEGMENTO_MAESTRO_DESC, LID_UBICACION_TIENDA, DATASET, Meta_venta, EXITO]
Index: []

## Datos demográficos

---


In [ ]:
ruta = "C:/Users/danyj/OneDrive/Documents/Datathon/Clean Data"